File: PythonClient/multirotor/ptz_camera.py

# AirSim PTZ Camera Simulator
#
# This script simulates a Pan-Tilt-Zoom (PTZ) camera using Microsoft AirSim.
# It performs two main tasks concurrently:
# 1. Streams video from an AirSim camera to a UDP port, encoded in H.264.
# 2. Listens for keyboard input to control the camera's pan, tilt, and zoom.
#
# Author: Gemini
# Date: August 2, 2025

# --- Installation ---
# You need to install the following Python packages:
# pip install airsim opencv-python numpy keyboard

# --- AirSim Setup ---
# 1. Make sure you have an AirSim environment (e.g., Blocks, Neighborhood) running.
# 2. Your AirSim settings.json file should be configured to enable the API.
#    A minimal settings.json looks like this:
#    {
#      "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/main/docs/settings.md",
#      "SettingsVersion": 1.2,
#      "SimMode": "Multirotor",
#      "Vehicles": {
#        "Drone1": {
#          "VehicleType": "SimpleFlight",
#          "X": 0, "Y": 0, "Z": 0
#        }
#      }
#    }

# --- How to Run ---
# 1. Start your AirSim environment.
# 2. Run this Python script from your terminal: python your_script_name.py
# 3. Open a network stream in a video player like VLC with the following URL:
#    udp://@:5000
# 4. Focus on the terminal window running the script and use the control keys.

import airsim
import cv2
import numpy as np
import threading
import time
import keyboard
import math

# --- Configuration ---
# Streaming Configuration
UDP_IP = "127.0.0.1"  # IP address to stream to (localhost)
UDP_PORT = 5000         # Port to stream to
IMG_WIDTH = 1280        # Video frame width
IMG_HEIGHT = 720        # Video frame height
FPS = 30                # Frames per second

# PTZ Control Configuration
VEHICLE_NAME = "Drone1" # Must match the name in your AirSim settings.json
CAMERA_NAME = "0"       # "0" is the front-center camera
PAN_SPEED = 2.0         # Degrees per step for pan
TILT_SPEED = 2.0        # Degrees per step for tilt
ZOOM_SPEED = 2.0        # Degrees per step for zoom (FoV change)
MIN_FOV = 15            # Minimum Field of View (max zoom)
MAX_FOV = 90            # Maximum Field of View (min zoom)

# --- Global State ---
# This dictionary will hold the current PTZ state, shared between threads.
ptz_state = {
    'pan': 0.0,   # Current pan in degrees
    'tilt': 0.0,  # Current tilt in degrees
    'fov': MAX_FOV, # Current Field of View
    'running': True # Flag to stop threads gracefully
}

def setup_video_stream():
    """
    Sets up the GStreamer pipeline for H.264 encoding and UDP streaming.
    This function creates an OpenCV VideoWriter object.
    """
    # GStreamer pipeline for encoding and streaming
    # appsrc:      Gets raw frames from this script
    # videoconvert: Converts color space
    # x264enc:     Encodes the stream to H.264. 'tune=zerolatency' is crucial for live streaming.
    # rtph264pay:  Creates RTP packets for H.264
    # udpsink:     Sends the packets over UDP
    gst_pipeline = (
        f"appsrc ! videoconvert ! x264enc tune=zerolatency bitrate=500 speed-preset=superfast ! "
        f"rtph264pay ! udpsink host={UDP_IP} port={UDP_PORT}"
    )
    
    video_writer = cv2.VideoWriter(gst_pipeline, cv2.CAP_GSTREAMER, 0, FPS, (IMG_WIDTH, IMG_HEIGHT), True)
    
    if not video_writer.isOpened():
        print("Error: Failed to open GStreamer pipeline. Make sure GStreamer is installed correctly.")
        return None
        
    print(f"✅ GStreamer pipeline initialized. Streaming to udp://{UDP_IP}:{UDP_PORT}")
    return video_writer

def video_streamer_thread(client, video_writer):
    """
    This function runs in a separate thread. It continuously captures images
    from AirSim, processes them, and writes them to the video stream.
    """
    print("🚀 Video streaming thread started.")
    
    # Define the request for a scene image
    request = airsim.ImageRequest(CAMERA_NAME, airsim.ImageType.Scene, False, False)

    while ptz_state['running']:
        start_time = time.time()
        
        # Get image from AirSim
        responses = client.simGetImages([request])
        response = responses[0]

        if response.image_data_uint8:
            # Convert the raw image data to a NumPy array
            img_1d = np.frombuffer(response.image_data_uint8, dtype=np.uint8)
            # Reshape the array to an image (3 channels for BGR)
            img_bgr = img_1d.reshape(response.height, response.width, 3)
            
            # Resize if necessary to match our streaming resolution
            if response.width != IMG_WIDTH or response.height != IMG_HEIGHT:
                img_bgr = cv2.resize(img_bgr, (IMG_WIDTH, IMG_HEIGHT))

            # Write the frame to the GStreamer pipeline
            video_writer.write(img_bgr)

        # Regulate the loop to approximate the desired FPS
        elapsed_time = time.time() - start_time
        sleep_time = (1.0 / FPS) - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time)
            
    print("🛑 Video streaming thread stopped.")
    video_writer.release()


def ptz_controller():
    """
    This function runs in the main thread. It listens for keyboard inputs
    and updates the shared ptz_state dictionary.
    """
    print("\n--- PTZ Camera Controls ---")
    print("  Pan Left:    'a' or Left Arrow")
    print("  Pan Right:   'd' or Right Arrow")
    print("  Tilt Up:     'w' or Up Arrow")
    print("  Tilt Down:   's' or Down Arrow")
    print("  Zoom In:     '+' or '='")
    print("  Zoom Out:    '-'")
    print("  Exit:        'esc'")
    print("---------------------------\n")
    print("ℹ️  Focus this terminal window to use controls.")

    while ptz_state['running']:
        # Pan controls
        if keyboard.is_pressed('a') or keyboard.is_pressed('left'):
            ptz_state['pan'] -= PAN_SPEED
        if keyboard.is_pressed('d') or keyboard.is_pressed('right'):
            ptz_state['pan'] += PAN_SPEED
            
        # Tilt controls
        if keyboard.is_pressed('w') or keyboard.is_pressed('up'):
            ptz_state['tilt'] -= TILT_SPEED
        if keyboard.is_pressed('s') or keyboard.is_pressed('down'):
            ptz_state['tilt'] += TILT_SPEED
            
        # Zoom controls
        if keyboard.is_pressed('+') or keyboard.is_pressed('='):
            ptz_state['fov'] -= ZOOM_SPEED
        if keyboard.is_pressed('-'):
            ptz_state['fov'] += ZOOM_SPEED
            
        # Exit control
        if keyboard.is_pressed('esc'):
            ptz_state['running'] = False
            print("\nESC pressed. Shutting down...")
            
        # Clamp values to their limits
        ptz_state['tilt'] = max(-89.0, min(89.0, ptz_state['tilt'])) # Limit tilt to avoid gimbal lock issues
        ptz_state['fov'] = max(MIN_FOV, min(MAX_FOV, ptz_state['fov']))
        ptz_state['pan'] %= 360 # Keep pan within 0-360 range

        time.sleep(0.05) # Small delay to prevent high CPU usage

def main():
    """
    Main function to initialize AirSim client, start threads,
    and apply PTZ updates.
    """
    # Initialize AirSim client
    client = airsim.MultirotorClient()
    try:
        client.confirmConnection()
        client.enableApiControl(True, VEHICLE_NAME)
        print("✅ Successfully connected to AirSim and enabled API control.")
    except Exception as e:
        print(f"Error: Could not connect to AirSim. Please ensure it is running.")
        print(f"Details: {e}")
        return

    # Setup video streaming
    video_writer = setup_video_stream()
    if not video_writer:
        client.enableApiControl(False, VEHICLE_NAME)
        return

    # Start the video streaming thread
    stream_thread = threading.Thread(target=video_streamer_thread, args=(client, video_writer))
    stream_thread.start()
    
    # Start the keyboard listener in the main thread
    # This function will block until ESC is pressed
    ptz_controller_thread = threading.Thread(target=ptz_controller)
    ptz_controller_thread.start()

    # Main loop to apply PTZ updates from the shared state
    last_pan, last_tilt, last_fov = -1, -1, -1
    
    while ptz_state['running']:
        pan, tilt, fov = ptz_state['pan'], ptz_state['tilt'], ptz_state['fov']
        
        # Only update AirSim if there's a change, to reduce API calls
        if pan != last_pan or tilt != last_tilt:
            # AirSim uses quaternions for orientation. We convert from Euler angles (pitch, roll, yaw).
            # Pan corresponds to Yaw, Tilt corresponds to Pitch. Roll is kept at 0.
            q = airsim.to_quaternion(math.radians(tilt), 0, math.radians(pan))
            client.simSetCameraOrientation(CAMERA_NAME, q, vehicle_name=VEHICLE_NAME)
            last_pan, last_tilt = pan, tilt
            
        if fov != last_fov:
            client.simSetCameraFov(CAMERA_NAME, fov, vehicle_name=VEHICLE_NAME)
            last_fov = fov
            
        time.sleep(0.02) # Update rate for applying camera changes

    # Cleanup
    print("Waiting for threads to join...")
    stream_thread.join()
    ptz_controller_thread.join()
    client.enableApiControl(False, VEHICLE_NAME)
    print("✅ Application has been shut down gracefully.")


if __name__ == "__main__":
    print(cv2.getBuildInformation())
    main()

File: PTZCamera/ptz_camera.cpp
// AirSim PTZ Camera Simulator (C++ - Refactored for Thread Safety)
//
// This application simulates a Pan-Tilt-Zoom (PTZ) camera using Microsoft AirSim.
// It uses a thread-safe wrapper for AirSim calls and separates PTZ control
// into a background thread to prevent video stutter.
//
// Author: Gemini
// Date: August 3, 2025

// --- Dependencies ---
// - AirSim C++ Client (msrpc-c_cxx)
// - OpenCV 4+ (MUST be compiled with GStreamer support)
// - GStreamer 1.0+ (runtime libraries)

// --- How to Build ---
// 1. Make sure all dependencies are installed.
// 2. Create a 'build' directory: mkdir build && cd build
// 3. Run CMake: cmake ..
// 4. Compile: make (or open the generated solution in Visual Studio and build)

// --- How to Run ---
// 1. Start your AirSim environment with PX4 SITL.
// 2. Run the compiled executable from your build directory.
// 3. Open the stream in VLC using the 'stream.sdp' file.
// 4. Make sure the 'AirSim C++ Feed' window is in focus to use PTZ controls.

#include <iostream>
#include <string>
#include <vector>
#include <thread>
#include <chrono>
#include <cmath>
#include <mutex>
#include <atomic>

// AirSim C++ Client
#include "rpc/rpc_error.h"
#include "vehicles/multirotor/api/MultirotorRpcLibClient.hpp"

// OpenCV
#include "opencv2/opencv.hpp"

// --- Configuration ---
const std::string UDP_IP = "127.0.0.1";
const int UDP_PORT = 5000;
const int IMG_WIDTH = 1280;
const int IMG_HEIGHT = 720;
const int FPS = 30;
const std::string VEHICLE_NAME = "PX4";
const std::string CAMERA_NAME = "0";
const float PAN_SPEED = 2.0f;
const float TILT_SPEED = 2.0f;
const float ZOOM_SPEED = 2.0f;
const float MIN_FOV = 15.0f;
const float MAX_FOV = 90.0f;

using AirSimRpcException = rpc::rpc_error;

// --- Thread-Safe AirSim Wrapper ---
class AirsimWrapper {
public:
    AirsimWrapper() {}

    bool connect() {
        try {
            client_.confirmConnection();
            std::cout << "✅ Successfully connected to AirSim." << std::endl;
            return true;
        } catch (AirSimRpcException& e) {
            std::cerr << "Error: Could not connect to AirSim." << std::endl;
            std::cerr << "Details: " << e.get_error().as<std::string>() << std::endl;
            return false;
        }
    }

    void updatePtz(float pan, float tilt, float fov) {
        std::lock_guard<std::mutex> lock(client_mutex_);
        try {
            msr::airlib::Pose vehicle_pose = client_.simGetVehiclePose(VEHICLE_NAME);
            msr::airlib::Quaternionr orientation_quat = msr::airlib::VectorMath::toQuaternion(
                msr::airlib::Utils::degreesToRadians(tilt), 0, msr::airlib::Utils::degreesToRadians(pan));
            
            // The camera's position is relative to the vehicle's frame, so (0,0,0) is correct here.
            msr::airlib::Pose new_camera_pose(Eigen::Vector3f(0, 0, 0), orientation_quat);
            
            client_.simSetCameraPose(CAMERA_NAME, new_camera_pose, VEHICLE_NAME);
            client_.simSetCameraFov(CAMERA_NAME, fov, VEHICLE_NAME);
        } catch (AirSimRpcException& e) {
            std::cerr << "Error updating camera: " << e.get_error().as<std::string>() << std::endl;
        }
    }

    std::vector<msr::airlib::ImageCaptureBase::ImageResponse> getImages() {
        std::lock_guard<std::mutex> lock(client_mutex_);
        const std::vector<msr::airlib::ImageCaptureBase::ImageRequest> request = {
            msr::airlib::ImageCaptureBase::ImageRequest(CAMERA_NAME, msr::airlib::ImageCaptureBase::ImageType::Scene, false, false)
        };
        return client_.simGetImages(request, VEHICLE_NAME);
    }

private:
    msr::airlib::MultirotorRpcLibClient client_;
    std::mutex client_mutex_;
};

// --- Shared State for Threads ---
std::atomic<bool> running(true);
std::atomic<float> pan(0.0f);
std::atomic<float> tilt(0.0f);
std::atomic<float> fov(MAX_FOV);

// --- PTZ Control Thread ---
// This thread sends PTZ updates to AirSim in the background.
void ptz_control_thread(AirsimWrapper& airsim_wrapper) {
    while (running) {
        // Read the latest atomic values
        float current_pan = pan.load();
        float current_tilt = tilt.load();
        float current_fov = fov.load();

        airsim_wrapper.updatePtz(current_pan, current_tilt, current_fov);
        
        // Send updates at a reasonable rate
        std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }
}

// --- Main Thread (Image Handling and Keyboard Input) ---
int main() {
    // --- GStreamer Pipeline Setup ---
    std::string gst_pipeline_str =
        "appsrc ! videoconvert ! "
        "x264enc tune=zerolatency bitrate=3000 speed-preset=fast ! "
        "rtph264pay ! "
        "udpsink host=" + UDP_IP + " port=" + std::to_string(UDP_PORT) + " sync=false buffer-size=1000000";

    cv::VideoWriter video_writer;
    video_writer.open(gst_pipeline_str, cv::CAP_GSTREAMER, 0, FPS, cv::Size(IMG_WIDTH, IMG_HEIGHT), true);
    if (!video_writer.isOpened()) {
        std::cerr << "Error: Failed to open GStreamer VideoWriter." << std::endl;
        return -1;
    }
    std::cout << "✅ GStreamer VideoWriter opened successfully." << std::endl;
    std::cout << "Streaming to udp://" << UDP_IP << ":" << UDP_PORT << std::endl;

    // --- AirSim Setup ---
    AirsimWrapper airsim_wrapper;
    if (!airsim_wrapper.connect()) {
        return -1;
    }
    
    // --- Start Background Thread for PTZ Control ---
    std::thread ptz_thread(ptz_control_thread, std::ref(airsim_wrapper));

    std::cout << "\n--- PTZ Camera Controls ---" << std::endl;
    std::cout << "  Pan Left:    'a'" << std::endl;
    std::cout << "  Pan Right:   'd'" << std::endl;
    std::cout << "  Tilt Up:     'w'" << std::endl;
    std::cout << "  Tilt Down:   's'" << std::endl;
    std::cout << "  Zoom In:     'e'" << std::endl;
    std::cout << "  Zoom Out:    'q'" << std::endl;
    std::cout << "  Exit:        'esc'" << std::endl;
    std::cout << "---------------------------\n" << std::endl;
    std::cout << "ℹ️  Focus the 'AirSim C++ Feed' window to use controls." << std::endl;

    // --- Main Loop for Image Acquisition, Display, and Input ---
    while (running) {
        auto start_time = std::chrono::high_resolution_clock::now();

        // Get image from AirSim
        auto response = airsim_wrapper.getImages();

        if (!response.empty() && !response[0].image_data_uint8.empty()) {
            cv::Mat img_bgr(response[0].height, response[0].width, CV_8UC3, (void*)response[0].image_data_uint8.data());
            if (!img_bgr.empty()) {
                cv::Mat resized_img;
                cv::resize(img_bgr, resized_img, cv::Size(IMG_WIDTH, IMG_HEIGHT));
                
                cv::imshow("AirSim C++ Feed", resized_img);
                video_writer.write(resized_img);
            }
        }

        // Handle keyboard input
        int key = cv::waitKey(1) & 0xFF;
        float current_pan = pan.load();
        float current_tilt = tilt.load();
        float current_fov = fov.load();

        switch (key) {
            case 'a': current_pan -= PAN_SPEED; break;
            case 'd': current_pan += PAN_SPEED; break;
            case 'w': current_tilt -= TILT_SPEED; break;
            case 's': current_tilt += TILT_SPEED; break;
            case 'e': current_fov -= ZOOM_SPEED; break;
            case 'q': current_fov += ZOOM_SPEED; break;
            case 27: running = false; break;
        }

        // Clamp and update atomic variables
        pan.store(fmod(current_pan, 360.0f));
        tilt.store(std::max(-89.0f, std::min(89.0f, current_tilt)));
        fov.store(std::max(MIN_FOV, std::min(MAX_FOV, current_fov)));

        // Regulate loop speed
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
        long sleep_ms = (1000 / FPS) - duration.count();
        if (sleep_ms > 0) {
            std::this_thread::sleep_for(std::chrono::milliseconds(sleep_ms));
        }
    }

    // --- Cleanup ---
    std::cout << "Cleaning up resources..." << std::endl;
    if (ptz_thread.joinable()) {
        ptz_thread.join();
    }
    video_writer.release();
    cv::destroyAllWindows();
    std::cout << "✅ Application has been shut down gracefully." << std::endl;

    return 0;
}
